main/src/library/Array.flix

```diff
@@ -785,6 +785,21 @@ namespace Array {
         };
         loop(length(arr) - 1, s)
 
+    ///
+    /// Applies `f` to a start value `z` and all elements in `a` going from right to left.
+    ///
+    /// That is, the result is of the form: `f(a[0], ...f(a[n-1], f(a[n], z))...)`.
+    /// A `foldRight` allows early termination by not calling the continuation.
+    ///
+    pub def foldRight(f: (a, Unit -> b & Impure) -> b & ef, z: b, arr: Array[a]): b & Impure =
```

Comment:
I'm not sure about if the purity is correct?

Reply:
I think the function `f` should have effect `ef`, but the overall function is still impure.

Reply:
It does not seem to like `f` having effect `ef`. Writing
```foldRight(f: (a, Unit -> b & ef) -> b & ef, z: b, arr: Array[a]): b & Impure```

i get the error
```
>> The type scheme: '∀(a, b). (a -> ((Unit ->{Impure} b) ->{Impure} b)) -> (b -> (Array[a] ->{Impure} result))' cannot be generalized to '∀(a, ef, b). (a -> ((Unit ->{ef} b) ->{ef} b)) -> (b -> (Array[a] ->{Impure} b))'.

10 |     pub def foldRight(f: (a, Unit -> b & ef) -> b & ef, z: b, arr: Array[a]): b & Impure =
                 ^^^^^^^^^
                 unable to generalize the type scheme.

The declared type does not match the inferred type:

  Declared: ∀(a, ef, b). (a -> ((Unit ->{ef} b) ->{ef} b)) -> (b -> (Array[a] ->{Impure} b))
  Inferred: ∀(a, b). (a -> ((Unit ->{Impure} b) ->{Impure} b)) -> (b -> (Array[a] ->{Impure} result))
```

but `foldRight(f: (a, Unit -> b & Impure) -> b & ef, z: b, arr: Array[a]): b & Impure` compiles

Reply:
OK, you are probably right, so lets go with that.